Refactor

timothee-haudebourg opened the pull request:
This PR aims at refactoring `ssi` to make better use of Rust traits.
- Abstract the credential/presentation types (using traits) to accept user defined types.
- Unify verifiable claim types (JWT, W3C VC). Have a uniform `Verifable<Claims>` type family for verifiable claims, independent of the underlying implementation.
- ~TreeLDR integration~ (although it was one of the main motivation for this refactor, integration with TreeLDR will be *de facto* possible using the new abstractions. `ssi` will *not* depend on TreeLDR).

This is a work in early progress.

# Status

Most of the work is completed. There are still some DID methods that needs to be reintroduced. Most of the remaining work is testing everything (and that's a lot).

Below are all the parts that needs to be added or changed in `ssi`:

- [x] Generate VC datamodel with TreeLDR
- [x] JWT VC (new dedicated `ssi-vc-jwt` library) 
  - [x] JWT+LD decoding
  - [x] verification
  - [x] JWT+LD signing
  - [x] encoding
- [x] Add an independent library for Verification Methods (`ssi-verification-methods` library), with VMs clearly defined.
  - [x] API design
  - [x] Verification methods implementation 
    - [x] `EcdsaSecp256k1RecoveryMethod2020`
    - [x] `EcdsaSecp256k1VerificationKey2019`
    - [x] `EcdsaSecp256r1VerificationKey2019`
    - [x] `Ed25519VerificationKey2018`
    - [x] `Ed25519VerificationKey2020`
    - [x] `JsonWebKey2020`
    - [x] `Multikey`
    - [x] `RsaVerificationKey2018`
    - [x] `Ed25519PublicKeyBLAKE2BDigestSize20Base58CheckEncoded2021`
    - [x] `P256PublicKeyBLAKE2BDigestSize20Base58CheckEncoded2021`
    - [x] `TezosMethod2021`
    - [x] `AleoMethod2021`
    - [x] `BlockchainVerificationMethod2021`
    - [x] `Eip712Method2021`
- [x] DIDs (`ssi-dids` library)
  - [x] Improve the `DIDResolver` trait
  - [x] Use the `ssi-verification-methods` library
  - [x] Reintroduce DID methods
    - [x] `tz` (Tezos)
    - [x]  `jwk`
    - [x] `pkh` (https://github.com/spruceid/ssi/pull/539)
    - [x] `key` (https://github.com/spruceid/ssi/pull/540)
    - [x] `web` (https://github.com/spruceid/ssi/pull/541)
    - [x] `ethr` (https://github.com/spruceid/ssi/pull/542)
    - [x] `ion` (https://github.com/spruceid/ssi/pull/543)
    - [x] ~`sol`~ (deprecate)
    - [x] ~`onion`~ (deprecate)
    - [x] ~`webkey`~ (deprecate)
    - [ ] `test` (test suite? not sure if required)
- [x] Data Integrity (in the `ssi-vc-ldp` library)
  - [x] VC decoding from JSON-LD
  - [x] VC encoding
  - [x] `CryptographicSuite` trait design 
  - [x] Cryptosuites implementations
    - [x] `EcdsaSecp256k1Signature2019`
    - [x] `EcdsaSecp256r1Signature2019`
    - [x] `Ed25519Signature2018`
    - [x] `Ed25519Signature2020`
    - [x] `eddsa-2022`
    - [x] `EthereumEip712Signature2021`
    - [x] `JsonWebSignature2020`
    - [x] `RsaSignature2018`
    - [x] `EcdsaSecp256k1RecoverySignature2020`
    - [x] `AleoSignature2021`
    - [x] `Eip712Signature2021`
    - [x] `EthereumPersonalSignature2021`
    - [x] `SolanaSignature2021`
  - [x] Any VC API
- [x] Testing

timothee-haudebourg commented on ssi-crypto/src/verification.rs:
```diff
@@ -0,0 +1,340 @@
+
+        /// Set of proof purposes.
+        #[derive(Debug, Default, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+        pub struct ProofPurposes {
```
The `ProofPurposes` type represents a set of `ProofPurpose` and expands to something like
```rust
struct ProofPurposes {
    pub assertion_method: bool,
    pub authentication: bool
    pub capability_invocation: bool,
    pub capability_delegation: bool,
    pub key_agreement: bool
}
```
Unlike a `Vec<ProofPurpose>` it does not require a memory allocation, it implements `Copy`, cannot include duplicates, provides a simple O(1) way to test included members (`p.assertion_method` instead of `p.contains(&ProofPurpose::AssertionMethod)`), and set operations (union, intersection, ...) using bit operators.

sbihel replied:
Ah right there can't be duplicates, but then why complicate things instead of having 
```rust
struct ProofPurposes {
  assertion_method: Option<AssertionMethod>,
...
}
```

Which I realise is basically `VerificationRelationships`? I guess I should just wait to see an example/test, but I'm afraid this is over complicated for minimal benefits

timothee-haudebourg replied:
You can find a full example that shows how to build `ProofPurposes` using bit or operations here: https://github.com/spruceid/ssi/blob/c90d8c209f8938f3bb2ea34b2c199fd5b8d802d4/ssi-vc-ldp/examples/sign.rs#L73

timothee-haudebourg commented on ssi-crypto/src/verification.rs:
```diff
@@ -0,0 +1,340 @@
+            }
+        }
+
+        impl BitOr for ProofPurposes {
```
Its for set operations. Here the union of `ProofPurposes` so you can write
```rust
my_proof_purposes | ProofPurpose::AssertionMethod
```

timothee-haudebourg commented on ssi-dids/src/did.rs:
```diff
@@ -0,0 +1,392 @@
+/// owned DIDs.
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[repr(transparent)]
+pub struct DID([u8]);
```
What is `bstr`? Or do you mean `str`? I'm used to using `[u8]` for pure ASCII strings because it simplifies many things (like indexing the string), but yes, I guess we can use `str` instead since it's a superset of ASCII.

sbihel replied:
I meant https://crates.io/crates/bstr

timothee-haudebourg replied:
We don't care about UTF-8 here. DIDs only use ASCII characters, which are encoded the same regardless of the unicode encoding. I could use `str`, but `[u8]` is simpler to interact with. Maybe we could use https://docs.rs/ascii/1.1.0/ascii/struct.AsciiStr.html to make it clearer but I'm not sure its really necessary.

timothee-haudebourg replied:
Oh I just realized it may not be ASCII or UTF-8 at all! There is an option in the header to disable base64 encoding so the payload can be raw bytes.

timothee-haudebourg replied:
We could do that, but a typical design pattern in rust is to use DSTs for this kind of "specialized" strings. Just like there is `str`/`String`, `Path`\`PathBuf`, `OsStr`\`OsString`, `BStr`\`BString` in the `bstr` crate you showed me and `AsciiStr`\`AsciiString` in the `ascii` crate. It avoids a lot of unnecessary memory allocations. Here I did the same with `DID`\`DIDBuf`.

The comment "owned DIDs" above the `DID` type is wrong, I removed it. It is not owned but borrowed.

timothee-haudebourg commented on ssi-dids/src/document.rs:
```diff
@@ -0,0 +1,292 @@
+    ///
+    /// See: <https://www.w3.org/TR/did-core/#did-controller>
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub controller: Option<OneOrMany<DIDBuf>>,
```
Yeah I think that's a good idea. Also, when `None` is semantically equivalent to `Some(empty_collection)`, I think we should remove the `Option` wrapper. It doesn't add anything.

timothee-haudebourg commented on ssi-dids/src/resolution/http.rs:
```diff
@@ -0,0 +1,146 @@
+/// A DID Resolver implementing a client for the [DID Resolution HTTP(S)
+/// Binding](https://w3c-ccg.github.io/did-resolution/#bindings-https).
+#[derive(Debug, Clone)]
+pub struct HTTPDIDResolver {
```
I agree, this could be moved to its own library.

timothee-haudebourg commented on ssi-jws/src/lib.rs:
```diff
@@ -3,17 +3,216 @@
 
+/// JWS in compact serialized form.
+#[repr(transparent)]
+pub struct CompactJWS([u8]);
```
I'm not changing the implementation itself, just adding a type to annotate serialized, compact JWS properly. I don't think `jose_jws` provides any type to represent compact JWS. We could replace our implementation of JWS with `json_jws` and still keep this type.

timothee-haudebourg commented on ssi-vc-ldp/src/suite.rs:
```diff
@@ -0,0 +1,184 @@
+}
+
+/// Built-in Data Integrity cryptographic suites types.
+pub enum SuiteType {
```
I see what you mean. It's not clear to me what would be the best way to handle unknown suites yet, but I'll try to add an `Unknown` variant and see if that feels natural.

timothee-haudebourg replied:
I see what you mean. It's not clear to me what would be the best way to handle unknown suites yet, but I'll try to add an Unknown variant and see if that feels natural.

timothee-haudebourg commented:
This is finally happening! Everything compiles, every test, doc test, example can run fine, no warnings, no broken doc links. I've fixed, tested and documented features. Now it is not perfect, there are a lot of ugly things living in `ssi_jwk`, `ssi_jws` for instance, but I'll consider it out of scope. Other issues are directly related to the refactor:

  - Before, the various verification functions would return a `VerificationResult` containing warnings and errors. There was no distinction between verification failure (the signature/proof is not valid), and verification errors (we could not verify the signature for some reason, it may or may not be valid). I tried to do the distinction using a more fine grain result type `Result<ProofValidity, VerificationError>` where `ProofValidity` captures the validity of the proof (it may be invalid) and `VerificationError` for when the verification process failed. I think its a good idea overall but
    - I haven't done it right everywhere. Some errors that should raise `ProofValidity::Invalid` instead throw a `VerificationError`
    - Ideally `ProofValidity::Invalid` would also gives the reason why the verification failed, but for now there is none
    - Warnings are lost, but we could add them to the `ProofValidity::Valid` variant
 - I've added a `Validate` trait to validate claims independently of the proof/signature. It can be used to check the expiration date of a credential for instance. I think this trait could be improved so we can pass an "environment" to the validation function. This is required for some advanced validation checks, such as validating the `aud` (audience) claim of a JWT against the application domain. To do it correctly the validation function needs to know the application domain, which could be passed through this "environment".

These are points we can discuss but I think overall the refactor is ready to be reviewed.

itsbalamurali commented:
@sbihel any ETA on merging this?

timothee-haudebourg commented:
`rustfmt` is failing because the `crates/multicodec/src/table.rs` file is not found. This file is generated by the `build.rs` script of `ssi_multicodec` from the `crates/multicodec/src/table.csv` file. I'll try to find a way to run the build script before `rustfmt`.

clehner commented on README.md:
```diff
@@ -3,45 +3,245 @@
+// public key used to sign the JWT.
+let vm_resolver = ExampleDIDResolver::default().with_default_options();
+
+assert!(vc.verify(&vm_resolver).await.expect("verification failed").is_valid());
```
snippet runs successfully as example script, with `examples` feature and wrapped in a main function

clehner commented on README.md:
````diff
@@ -3,45 +3,245 @@
+
+
+```rust
+use json_syntax::json;
````
```suggestion
```
compile warning "unused import"
snippet runs successfully anyway

clehner commented on crates/claims/crates/data-integrity/src/any/suite.rs:
```diff
@@ -0,0 +1,618 @@
+
+impl AnySuite {
+    #[cfg(feature = "eip712")]
+    pub fn requires_eip721(&self) -> bool {
```
```suggestion
    pub fn requires_eip712(&self) -> bool {
```
& below and in `crates/claims/crates/data-integrity/src/any/options.rs`

timothee-haudebourg replied:
Fixed.

timothee-haudebourg commented on crates/claims/crates/vc/src/data_model/credential.rs:
```diff
@@ -0,0 +1,224 @@
+    ///
+    /// Validation may fail even if the credential proof is successfully
+    /// verified.
+    fn is_valid_credential(&self) -> bool {
```
I agree, validation deserves a bit more work, but should I refine it in this PR? For instance, besides a proper `Result` output type, it would be nice to have an input argument providing some kind of environment used to validate some claims such as the expiration date (requires the current date), the intended audience (requires the verifier's domain name), etc.

sbihel replied:
Yeah, and also this is where the distinction between validation and verification becomes fuzzy. Considering making an error type that would work for all credential would be very difficult, I would say maybe change to `fn validate(&self) -> Result<(), String>` in this PR and that should be good enough for a while

timothee-haudebourg commented on crates/claims/crates/data-integrity/core/src/suite.rs:
```diff
@@ -0,0 +1,326 @@
+    TooLong,
+
+    #[error("invalid message: {0}")]
+    InvalidMessage(Box<dyn 'static + std::error::Error>),
```
I'll make it a `String`, I agree this much complexity is not required.

sbihel replied:
Sounds good, also just to clarify when I say "they're not actually useful" it's that the crypto libraries purposefully don't return anything helpful to avoid leaking information

timothee-haudebourg replied:
I've put a `String` instead, and I also "standardized" all the error types so we have a single error type for signing errors, same for signature/proof validation. Also made sure none of those error types have type parameters because that's annoying to deal with.

ianhamilton87 commented:
@timothee-haudebourg @clehner @sbihel - any ETA on the completion of this PR?
Reason I ask is because merging this PR is blocking one of mine: https://github.com/spruceid/ssi/pull/545

Thanks.

sbihel commented:
There's is no ETA but we're hoping to get it merged and published within two months.

timothee-haudebourg commented on crates/dids/core/src/method_resolver.rs:
```diff
@@ -0,0 +1,132 @@
+
+    async fn resolve_verification_method<'a, 'm: 'a>(
+        &'a self,
+        _issuer: Option<&'a iref::Iri>,
```
Some implementations may rely on the issuer to fetch a verification method (some "default issuer verification method"). The only implementation we provide is based on DIDs, where there is no such notion of "default issuer verification method". So its true we don't use it ourselves, is it enough of a reason to remove it?

timothee-haudebourg commented on crates/dids/core/src/method_resolver.rs:
```diff
@@ -0,0 +1,132 @@
+    async fn resolve_verification_method<'a, 'm: 'a>(
+        &'a self,
+        _issuer: Option<&'a iref::Iri>,
+        method: Option<ReferenceOrOwnedRef<'m, M>>,
```
The `method` argument you mean? If the method is not present some resolver may rely on the issue to find a default one.

timothee-haudebourg commented:
I've made some changes to follow the reviews. With a few days away from ssi I had some fresh eyes, so I've fixed some other issues that were necessary in my opinion:
  - Normalized all the claim signature/verification error types. There is now one `SignatureError` and one `ProofValidationError` used across all the `sign`/`verify` functions, whatever the claim type (Data-Integrity, JWS, JWT, etc.).
  - Provided an environment to the claim validation method `Validate::validate`. This can be used to deterministically set the time of verification for instance, used to check the expiration/issuance date claims for instance. The function now returns a result, giving an insight on why the verification failed (if the claim expired for instance).
  - Normalized the verification/signature API across claim types. Now besides Data-Integrity claims, the same verification pipeline is (can be) used to verify JWS, JWT, etc. I've added some example in the crate-level documentation of `ssi-jws` and `ssi-jwt`. I've also added documentation of this "verification pipeline" here: [crates/claims/core/src/verification/mod.rs](https://github.com/spruceid/ssi/blob/9a0af26e2219ed27087cd86a1a04bc2c3b237469/crates/claims/core/src/verification/mod.rs) (maybe not the best place since this won't appear in the generated doc).
  - Changed the way we encode JWT registered claims. It is now very similar to `cose-rs`, which we will probably use in the future to verify CBOR-encoded claims. It is also more memory efficient.
  - Completely separated JWS/JWT from the `ssi-verification-methods` crate, just so it stays reasonable to use the `ssi-jws`/`ss-jwt` crates alone, without the rest of `ssi`.

sbihel commented:
There are still some error types that are missing `Send` and would probably benefit from moving to a `String` or `anyhow::Error` -- at least `MessageSignatureError `. Have you tried compiling the VC API in DIDKit in https://github.com/spruceid/didkit/pull/384?

jceb commented:
🎉
